### PR TITLE
[PyTorch] Add MaybeOwned::operator*() &&

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned
* #55316 [PyTorch] Add & use inferExpandGeometry_dimvector
* #55247 [PyTorch][easy] Use sizes()[x] instead of size(x) in addr
* #55246 [PyTorch] Update expand_size API to match expand_inplace
* #55245 [PyTorch] Update expand_outplace API to match expand_inplace
* **#55244 [PyTorch] Add MaybeOwned::operator*() &&**
* #55180 [PyTorch] Use infer_size_dimvector in ExpandUtils
* #55065 [PyTorch] New expand_inplace API with MaybeOwned<Tensor> and no unary tuples

Add the ability to move from the underlying object in a `MaybeOwned`.

FWIW, `MaybeOwned` is new territory for me personally and this move-and-dereference operation is even more so, but I think it makes sense and the tests pass.

Differential Revision: [D27522809](https://our.internmc.facebook.com/intern/diff/D27522809/)

### CODE REVIEW

**Comment by swolchok** [comment]
ping

**Comment by swolchok** [comment]
ping -- it's been at least a day with no review

File: c10/util/MaybeOwned.h [hunk 1]
```diff
@@ -103,6 +103,19 @@ class MaybeOwned final {
...
     return isBorrowed_ ? borrow_ : &own_;
   }
+
+  // If borrowed, copy the underlying T. If owned, move from
+  // it. borrowed/owned state remains the same, and either we
+  // reference the same borrow as before or we are an owned moved-from
+  // T.
+  T operator*() && {
+    if (isBorrowed_) {
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(borrow_ != nullptr);
```
Comment (bhosmer) [inline c10/util/MaybeOwned.h:113]: No action needed, but curious the motivation behind checking this on derefs rather than on construction/assignment

**Review by bhosmer — APPROVED** [review]

File: c10/util/MaybeOwned.h (hunk 1 above)
Comment (swolchok) [reply c10/util/MaybeOwned.h:113]: Default construction creates a null borrowed MaybeOwned to simplify initialization inside an if statement.

**Pull request closed without merging**